custom_components/ramses_cc/broker.py

```diff
@@ -170,18 +189,96 @@ async def async_save_client_state(self, *args, **kwargs) -> None:
             }
         )
 
+    def async_register_platform(
+        self,
+        platform: EntityPlatform,
+        add_new_devices: Callable[[RamsesRFEntity], None],
+    ) -> None:
+        """Register a platform for device addition."""
+        self.entry.async_on_unload(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_NEW_DEVICES.format(platform.domain), add_new_devices
+            )
+        )
+
+    async def _async_setup_platform(self, platform: str) -> None:
+        """Set up a platform."""
+        if platform not in self._platform_setup_tasks:
+            self._platform_setup_tasks[platform] = self.hass.async_create_task(
+                self.hass.config_entries.async_forward_entry_setup(self.entry, platform)
+            )
+        await self._platform_setup_tasks[platform]
+
+    async def async_unload_platforms(self) -> bool:
+        """Unload platforms."""
+        tasks = []
+
+        for platform, task in self._platform_setup_tasks.items():
+            if task.done():
+                tasks.append(
+                    self.hass.config_entries.async_forward_entry_unload(
+                        self.entry, platform
+                    )
+                )
+            else:
+                task.cancel()
+                tasks.append(task)
+
+        return all(await asyncio.gather(*tasks))
+
+    def _update_device(self, device: RamsesRFEntity) -> None:
+        if hasattr(device, "name") and device.name:
+            name = device.name
+        elif isinstance(device, Evohome):
+            name = "Controller"
```

Comment:
This is the simplest solution to maintain the current default name/ID of the main climate entity as `Controller` and `climate.controller`.

It does mean that anyone with multiple evohome systems will have two devices called `Controller` with entity IDs `climate.controller` and `climate.controller_2` but that's not really any different to the current arrangement. The device ID is available as the serial ID to disambiguate and the presence of unique IDs allows for custom renaming of the device name and entity name/ID as desired. I also imagine that this pretty rare!

Without this special case the device would be named after the slug and ID (so `evohome 01:123456`) which isn't really as friendly and would also require overrides at the entity level to maintain a name/ID of `Controller` and `climate.controller` so I chose not to go that route.

Reply:
> I also imagine that this pretty rare!

Not from the number of people who PM me to tell me it isn't!

> Without this special case the device would be ... so I chose not to go that route.

There are will breaking changes anyway... So just bite the bullet?

Actually, the best solution for them/us, I think, would be the same as here: [evohome wiki](https://github.com/zxdavb/evohome-async/wiki/Hass:-Multiple-Evohome-Locations): multiple integrations (i.e. add in **ramses_c1**, **ramses_c2**, etc.).

But they'll need multiple dongles in the short term, although I can see multiple ways of getting two **ramses_cc**s talking to one dongle.

Reply:
If we're happy with some change in behavior then the decision boils down to what the preferred name for the system device (and climate entity) is. Do we want it to be "Controller" or "evohome 01:123456"? I can see pros/cons to each.

Reply:
I'd say "controller 01:123456".

The unique ID must be be unique - everything else is forgivable.

My fear is, that I didn't set it up right - now is the time to sort that out.

Of note: this doesn't look right to me:

```python
class RamsesEntity(Entity):
    def __init__(self, ...) -> None:
        self._attr_unique_id = device.id
```

A whole bunch of entities will have the same unique-id. I think HA then does some woo to pick up the pieces... ?using the entity-id?

For example, in the entity registry:
```
        "entity_id": "binary_sensor.01_145038_01_window_open",
        "unique_id": "01:145038_01-window_open",
```
... for the controller, and one of the zones:
```json
        "entity_id": "climate.controller",
        "unique_id": "01:145038",
...
        "entity_id": "climate.ramses_cc_01_145038_00",
        "unique_id": "01:145038_00",
```

Reply:
The sensors handle that [here](https://github.com/zxdavb/ramses_cc/blob/1ff957a7c7565e7b52ee2d8fa722eca3744243b6/custom_components/ramses_cc/sensor.py#L157) and [here](https://github.com/zxdavb/ramses_cc/blob/1ff957a7c7565e7b52ee2d8fa722eca3744243b6/custom_components/ramses_cc/binary_sensor.py#L119). The RamsesEntity default is used for climate, water heater and remote entities which works out as only one entity from any one of those platform domains will be possible per device at present.

"Controller 01:123456" seems reasonable. That's probably clearer than using the slug.

Reply:
> That's probably clearer than using the slug.

The slugs are a little messy.

They are used for device _class_ (e.g. BDR, OTB, TRV (and sometimes device _type_: TRV vs TR2)) and also the CH/DHW _roles_ (TCS, ELE, RAD).

I would like to switch away from using slugs for roles... For example, is DHW a CS92A or the HW part of the CH/DHW system.
 ... and people always confuse the TCS (CH/DHW controller) with the CTL (CH/DHW controller!).

Reply:
I guess it would be nice to provide a friendlier name for all device types really. I used the slug for a prefix as it seemed like the best bet at the moment but it seemed worth considering the controller immediately as it's a bit more key.

Reply:
Adjusted to be "Controller 01:123456" for `Evohome` instances. Would it make sense to extend this to all `System` instances?

Reply:
> Adjusted to be "Controller 01:123456" for Evohome instances. Would it make sense to extend this to all System instances?

Yes, although I am not sure how many non-evohome systems there are.

Reply:
I want to change the entity_id of DHW to `water_heater.01_123456_hw` to match that of the climate entites, say: `climate.01_123456_01` - what is the best way to do that for config flow?

The issue with DHW is that - unlike zones - it has a name when instantiated.

So, I want to set the name of the evohome TCS/Zone/DHW regardless of name.

Reply:
Updated accordingly.

Reply:
> I want to change the entity_id of DHW to `water_heater.01_123456_hw` to match that of the climate entites, say: `climate.01_123456_01` - what is the best way to do that for config flow?
> 
> The issue with DHW is that - unlike zones - it has a name when instantiated.
> 
> So, I want to set the name of the evohome TCS/Zone/DHW regardless of name.

The best thing would be to explicitly provide a preferred `entity_id` in the same way as is [done for the sensors](https://github.com/zxdavb/ramses_cc/blob/94ff40e2d061950e22855ed14c0ec7f75bab82fc/custom_components/ramses_cc/sensor.py#L160). At present as you say the entity ID is generated from the entity name (inherited from the device in this case).

I assume the idea is that the entity ID will remain fixed at a known stable value regardless of zone renames etc.

Reply:
If the unique_id doesn't change, then the entity_id won't change, even if it (i.e. the 'offer' from the integration) is changed in the integration - you have to remove the entity from the registry.

Reply:
OK, Climate, Remote & WaterHeater all have:
```python
self.entity_id = f"{PLATFORM}.{device.id}"
```

... and the sensors remain:
```python
self._attr_unique_id = f"{device.id}-{entity_description.key}"
```

Reply:
Yep. I wasn't being very clear there. What I meant was that it would perhaps be wise to provide an explicit entity_id 'offer' so that it is not closely coupled to the device/entity name provided. ie. you could change the generated friendly name of the water heater or controller devices in a future version but still retain the same entity id and thus not affect expectations for automations/documentation etc. for newly created entries.

Reply:
I did also kind of wonder if an entity_id of `climate.01_123456` might make more sense for the controller to fit with the general approach (it's currently implicitly generated from the provided name as `climate.controller_01_123456`). Or, alternatively `climate.01_123456_controller`.

Reply:
> OK, Climate, Remote & WaterHeater all have:
> 
> ```python
> self.entity_id = f"{PLATFORM}.{device.id}"
> ```
> 
> ... and the sensors remain:
> 
> ```python
> self.entity_id = ENTITY_ID_FORMAT.format(f"{device.id}_{entity_description.key}")
> ```

That seems fairly sane to me. Although, for climate for example you can do:
```
from homeassistant.components.climate import ENTITY_ID_FORMAT
...
self.entity_id = ENTITY_ID_FORMAT.format(device.id)
```
Which gets you the same thing but seemed neater to me.

Reply:
I was making the change at the same time and also corrected the domain for remote and sensor entities which I'd botched. It's interesting that that gets corrected by HA anyway but much better if it's right to start with!

I've merged your changes in to resolve the conflicts.